### ISSUE TITLE
Syntactic Sugar for actions

### ISSUE THREAD

**Issue by cwohlman** [usefulio/pipes#3]
You should be able to specify an action in any of these ways:

`pipe.on('default', 'before', 'pre-process')`
`pipe.on('default', 'before', ['pre-process', 'other-action'])`
`pipe.on('default', 'before', 'pre-process', 'other-action')`
`pipe.on('default', 'before', 'pre-process', function (data) { console.log(data); })`

**Comment by gevou** [comment]
`pipe.on('default', 'before', ['pre-process', 'other-action'])`
`pipe.on('default', 'before', 'pre-process', 'other-action')`

Is the intention here to accept an actions-sequence, either as an array or as a set of N individual attributes?

**Comment by gevou** [comment]
Also, I understand that the idea here is to have named actions/functions.

Consider a named action that is initialized as: 
`pipe.on('default', 'before', 'pre-process', function (data) { console.log(data); })`

Should the user be able, after the initial definition, to assign it to other Cues or stages e.g.
`pipe.on('default', 'after', 'pre-process');`
`pipe.on('otherCue', 'before', 'pre-process');`
, or should it be somehow limited to either specific cue/stage (I cannot think of a reason though)?

ps. alternatively we can create a separate function specifically for declaring named actions.
`pipe.action('actionName', function (data) {...});'
which can then be used with`pipe.on`. The only benefit I see is clearer distinction between named-unnamed actions in the code. Any thoughts on that?

**Comment by cwohlman** [comment]
> Is the intention here to accept an actions-sequence, either as an array or as a set of N individual attributes?

yes

> Consider a named action that is initialized as: 
> pipe.on('default', 'before', 'pre-process', function (data) { console.log(data); })

The name of this action is `default` and it can be reused by reference:

`pipe.on('special', 'before', 'default')` (runs the default action before running the 'special' action.

**Comment by gevou** [comment]
The usage is currently as follows: 
step 1: initialize an instance of Pipe and declare the different execution stages that are shared across all named pipes, e.g. "before", "action", "after", using the constructor 
`var mypipe = Pipe(["before", "action", "after"]);`

step 2: use method `Pipe.on` to assign one or more actions to a specific stage and named pipe e.g.:
`mypipe.on('default', 'before', function (data) { console.log(data); })`
`mypipe.on('default', 'after', function (data) { console.log(data); })`
here the pipe name is `default` and the function that is assigned to the `before` stage will run before the one that is assigned to the `after` stage. 

step 3: use method `Pipe.do` to run a specific pipe with an `options` attribute that is passed to the actions.
`mypipe.do('default', options);`

The syntax you are using above is very different so I need to understand what is the intended usage, beyond declaring named/reusable actions which is clear.
a) to define and execute series of actions (either by reference or by assignment) on the fly without assigning to pipes/stages?
b) to be able to define the execution sequence when >1 actions exist in a single `.on` method call by using keywords? (Since, however, the execution sequence is left to right, what is the purpose of including a keyword like `before`?)
c) other?

**Comment by cwohlman** [comment]
We want to support pipe aliases:

`pipe.on('send', 'action', 'mailer-send')` This adds an `action` to the `send` pipe, the action which is added is an alias for the `mailer-send` action, and is equivalent to this: `pipe.on('send', 'action', function (options) { return this.do('mailer-send', options); }`

We also want to make it easier for users to add multiple actions at a time, so we allow them to do this:

`pipe.on('send', 'action', 'mailer-send', 'mailer-stamp')` which executes the `mailer-send` and `mailer-stamp` actions during the `action` stage of the `send` pipe.

In other words we want to support passing multiple arguments, or an array, instead of repeatedly calling `on`.

**Comment by gevou** [comment]
ok thanks. That's clear. 
In your example: `pipe.on('send', 'action', 'mailer-send')`
How do you initially assign the `mailer-send` alias to the respective action/function? 
for example: 

Inside an `.on` method call: 
`pipe.on('send', 'action', 'mailer-send', function (options) {...});` 
(the `function (options)` is aliased as `mailer-send` and automatically cued to `send`/`action` but you can also reference this elsewhere after this method call.) 

or using a dedicated action-alias assignment method? e.g. 
`pipe.action('mailer-send', function (options) {...});`
`pipe.on('send', 'action', 'mailer-send')`

**Comment by gevou** [comment]
(after re-reading your comments... ) 
Is the intention for a named `pipe` created using `.on` to also be used as an alias? e.g.
`pipe.on('mailer-send', 'action', function (options) {...});`
`pipe.on('aggregateSendPipe', 'action', 'mailer-send', 'other-transport-send', function() .... );`

**Comment by cwohlman** [comment]
Sorry for my late response, I missed the notification. Your latest comment is correct, every pipe you create is also a named action which can be used as part of other pipes.

**Comment by gevou** [comment]
ok

**Comment by gevou** [comment]
in the following case: 
`pipe.on('mailer-send', 'before', function (options) {...}); //action1`
`pipe.on('mailer-send', 'action', function (options) {...}); //action2`

`pipe.on('aggregatePipe', 'action', 'mailer-send');`

`pipe.do('aggregatePipe', options);`
Should this call both action1 (stage: before) and action2 (stage: action) of `mailer-send` or only action2?

**Comment by cwohlman** [comment]
Both.
